### PR TITLE
Collections that inherit from dict should should override copy()

### PR DESCRIPTION
**Pull request by chadrik**
Solves issue #1642.  A previous attempt at this, #1656, added `__copy__` but omitted `copy`, and it did not properly use self-type.  

One question regarding the previous PR: should the `__copy__` method be in the stubs at all?   As far as I can tell, `OrderedDict` does not have a `__copy__` method.  Was adding this a mistake or is it necessary for structural type checking of some sort?

### CODE REVIEW

File: stdlib/2/collections.pyi [hunk 1]
```diff
@@ -93,10 +93,13 @@ class Counter(Dict[_T, int], Generic[_T]):
     def __iand__(self, other: Counter[_T]) -> Counter[_T]: ...
     def __ior__(self, other: Counter[_T]) -> Counter[_T]: ...
 
+_OrderedDictT = TypeVar('_OrderedDictT', bound=OrderedDict)
```
Comment (ilevkivskyi) [inline stdlib/2/collections.pyi]: Personally, I prefer shorter names for type variables (also this is what PEP 8 currently recommends).

File: stdlib/2/collections.pyi [hunk 2]
```diff
@@ -93,10 +93,13 @@ class Counter(Dict[_T, int], Generic[_T]):
...
+
 class OrderedDict(Dict[_KT, _VT], Reversible[_KT], Generic[_KT, _VT]):
     def popitem(self, last: bool = ...) -> Tuple[_KT, _VT]: ...
     def __reversed__(self) -> Iterator[_KT]: ...
-    def __copy__(self) -> OrderedDict[_KT, _VT]: ...
+    def __copy__(self: _OrderedDictT) -> _OrderedDictT: ...
```
Comment (ilevkivskyi) [inline stdlib/2/collections.pyi]: There is actually no `Copyable` protocol in typing, so I would say it is safe to remove it, if it is not defined at runtime.

File: stdlib/2/collections.pyi [hunk 3]
```diff
@@ -93,10 +93,13 @@ class Counter(Dict[_T, int], Generic[_T]):
...
+    def copy(self: _OrderedDictT) -> _OrderedDictT: ...
```
Comment (ilevkivskyi) [inline stdlib/2/collections.pyi]: Note that there are some bugs with self-types in generic types (in mypy). But hopefully most of them will be fixed soon (as soon as I will have time to finish a patch I am working on).

File: stdlib/2/collections.pyi (hunk 1 above)
Comment (chadrik) [reply stdlib/2/collections.pyi]: I followed the convention already established in the file by `_UserDictT`, `_UserListT`, and `_UserStringT`.  I think it makes sense to keep it consistent.

File: stdlib/2/collections.pyi (hunk 2 above)
Comment (chadrik) [reply stdlib/2/collections.pyi]: will do.

**Comment by chadrik** [comment]
I guess I should add `Counter` too, but I'll wait for your go-ahead.

File: stdlib/2/collections.pyi (hunk 1 above)
Comment (chadrik) [reply stdlib/2/collections.pyi]: If you really hate it as is I can do `_OdictT`.  your call.

**Comment by ilevkivskyi** [comment]
You can add other copy methods in this PR using the current local `TypeVar` convention, but then please open a separate issue/PR on top of this to align this file with PEP8 (this also appears in some other files, so this requires a wider discussion/exposure).

**Comment by chadrik** [comment]
> You can add other copy methods in this PR using the current local TypeVar convention, 

Will do.

> but then please open a separate issue/PR on top of this to align this file with PEP 8 (this also appears in some other files, so this requires a wider discussion/exposure).

I'm probably the wrong guy to head up that PR because I personally find short TypeVar names to be too cryptic, and I independently adopted the same convention used in collections.pyi (especially for self-types).  I've always been taught to favor clarity over brevity.

**Comment by JelleZijlstra** [comment]
For what it's worth, I think slightly longer TypeVar names are fine here.

**Comment by chadrik** [comment]
This PR now includes `copy()` methods for `OrderedDict`, `defaultdict`, and `Counter`.

**Pull request merged**